File: lib/android/src/main/java/com/airbnb/android/react/navigation/TabViewManager.java
package com.airbnb.android.react.navigation;

import android.util.Log;

import com.facebook.react.bridge.ReadableMap;
import com.facebook.react.uimanager.SimpleViewManager;
import com.facebook.react.uimanager.ThemedReactContext;
import com.facebook.react.uimanager.annotations.ReactProp;

public class TabViewManager extends SimpleViewManager<TabView> {

    private static final String TAG = "TabViewManager";

    @Override
    public String getName() {
        return "NativeNavigationTabView";
    }

    @Override
    protected TabView createViewInstance(ThemedReactContext reactContext) {
        Log.d(TAG, "createViewInstance");
        return new TabView(reactContext, null);
    }

    @ReactProp(name = "route")
    public void setRoute(TabView view, String route) {
        Log.d(TAG, "setRoute");
        view.setRoute(route);
    }

    @ReactProp(name = "props")
    public void setProps(TabView view, ReadableMap props) {
        Log.d(TAG, "setProps");
        view.setProps(props);
    }

    @ReactProp(name = "config")
    public void setConfig(TabView view, ReadableMap config) {
        Log.d(TAG, "setConfig");
        view.setConfig(config);
    }
}

File: lib/ios/native-navigation/TabViewManager.swift
//
//  TabViewManager.swift
//  NativeNavigation
//
//  Created by Leland Richardson on 8/10/16.
//  Copyright © 2016 Airbnb. All rights reserved.
//

import UIKit;
#if !NN_NO_COCOAPODS
    import React
#endif

// MARK: - TabView

final class TabView: UIView {

//  init(implementation: ReactNavigationImplementation) {
//    super.init()
//    self.implementation = implementation
//  }

//  required init?(coder aDecoder: NSCoder) {
//    fatalError("init(coder:) has not been implemented")
//  }

  // MARK: Internal

  func setRoute(_ route: String!) {
    self.route = route
  }

  func setConfig(_ config: [String: AnyObject]) {
    self.prevConfig = self.renderedConfig
    self.renderedConfig = config
    implementation.reconcileTabConfig(
      tabBarItem: tabBarItem,
      prev: prevConfig,
      next: renderedConfig
    );
  }

  func setProps(_ props: [String: AnyObject]) {
    self.props = props
  }

  func getViewController() -> UIViewController? {

    if let viewController = viewController {
      return viewController
    }

    // TODO(lmr): handle non-RN tabs
    if let route = route {
      let vc = ReactViewController(moduleName: route, props: props).prepareViewControllerForPresenting()
      vc.tabBarItem = tabBarItem
      viewController = vc
    }

    return viewController

  }

  // MARK: Private

  private var implementation: ReactNavigationImplementation = ReactNavigationCoordinator.sharedInstance.navigation
  private var viewController: UIViewController?
  private var tabBarItem: UITabBarItem = UITabBarItem()
  private var route: String?
  private var props: [String: AnyObject] = [:]
  private var prevConfig: [String: AnyObject] = [:]
  private var renderedConfig: [String: AnyObject] = [:]

}

// MARK: - TabViewManager

private let VERSION: Int = 1

@objc(TabViewManager)
final class TabViewManager: RCTViewManager {
  override func view() -> UIView! {
    return TabView()
//    return TabView(implementation: ReactNavigationCoordinator.sharedInstance.navigation)
  }
}
